r/src/tpch-gen.cpp

```diff
@@ -0,0 +1,167 @@
+  }
+
+  arrow::AsyncGenerator<arrow::util::optional<compute::ExecBatch>> sink_gen;
+
+  MakeExecNodeOrStop("sink", plan.get(), {table}, compute::SinkNodeOptions{&sink_gen});
+
+  StopIfNotOk(plan->Validate());
+  StopIfNotOk(plan->StartProducing());
+
+  // If the generator is destroyed before being completely drained, inform plan
+  std::shared_ptr<void> stop_producing{nullptr, [plan](...) {
+                                         bool not_finished_yet =
+                                             plan->finished().TryAddCallback([&plan] {
+                                               return [plan](const arrow::Status&) {};
+                                             });
+
+                                         if (not_finished_yet) {
+                                           plan->StopProducing();
+                                         }
+                                       }};
+
+  return compute::MakeGeneratorReader(
+      table->output_schema(), [stop_producing, plan, sink_gen] { return sink_gen(); },
+      gc_memory_pool());
+}
+
+void Queue_Write_One_Table(const std::shared_ptr<compute::ExecPlan>& plan,
+                           std::string table_name, compute::ExecNode* table,
+                           std::string base_path,
+                           const std::shared_ptr<fs::FileSystem>& filesystem) {
+  auto format = std::make_shared<ds::ParquetFileFormat>();
+
+  auto partitioning_factory = arrow::dataset::HivePartitioning::MakeFactory();
+  std::vector<std::shared_ptr<arrow::Field>> fields;
+  auto partitioning = ValueOrStop(partitioning_factory->Finish(arrow::schema(fields)));
+
+  ds::FileSystemDatasetWriteOptions write_options;
+  write_options.file_write_options = format->DefaultWriteOptions();
+  write_options.existing_data_behavior =
+      ds::ExistingDataBehavior::kDeleteMatchingPartitions;
+  write_options.filesystem = filesystem;
+  write_options.base_dir = base_path + "/" + table_name;
+  write_options.partitioning = partitioning;
+  write_options.basename_template = "data-{i}.parquet";
+  write_options.max_partitions = 1024;
+
+  const ds::WriteNodeOptions options =
+      ds::WriteNodeOptions{write_options, table->output_schema()};
+
+  MakeExecNodeOrStop("write", plan.get(), {table}, options);
+}
+
+// [[arrow::export]]
+void Tpch_Dbgen_Write(const std::shared_ptr<compute::ExecPlan>& plan, double scale_factor,
+                      const std::shared_ptr<fs::FileSystem>& filesystem,
+                      std::string base_dir, std::string folder_name) {
+  arrow::dataset::internal::Initialize();
+
+  auto gen =
+      ValueOrStop(arrow::compute::internal::TpchGen::Make(plan.get(), scale_factor, 1000000));
+
+  auto base_path = base_dir + folder_name;
+  (void)filesystem->CreateDir(base_path);
+
+  static std::string tables[] = {"part",   "supplier", "partsupp", "customer",
+                                 "nation", "lineitem", "region",   "orders"};
+  for (auto& table_name : tables) {
```

Comment:
For this for loop on the other hand, we can do something a little more interesting: We can make an array of function pointers and just iterate through that. 

```
using TableNodeFn = Result<ExecNode*> (TpchGen::*)(std::vector<std::string>);
constexpr int kNumTables = 8;
std::array<TableNodeFn, kNumTables> tables = {
      &TpchGen::Supplier, &TpchGen::Part,     &TpchGen::PartSupp, &TpchGen::Customer,
      &TpchGen::Orders,   &TpchGen::Lineitem, &TpchGen::Nation,   &TpchGen::Region,
  };
for(auto fn : tables)
{
    compute::ExecNode *table = ValueOrStop((gen->*fn)());
    Queue_Write_One_Table(...);
}
```